File: src/examples/python/ir.py
#!/usr/bin/env python2
## -*- coding: utf-8 -*-
##
## Output
##
##  $ python ir.py
##  400000: mov rax, qword ptr [rip + 0x13b8]
##          ref!0 = (concat ((_ extract 7 0) (_ bv0 8)) ((_ extract 7 0) (_ bv49 8)) ((_ extract 7 0) (_ bv0 8)) ((_ extract 7 0) (_ bv50 8)) ((_ extract 7 0) (_ bv0 8)) ((_ extract 7 0) (_ bv51 8)) ((_ extract 7 0) (_ bv0 8)) ((_ extract 7 0) (_ bv52 8))) ; MOV operation
##          ref!1 = (_ bv4194311 64) ; Program Counter
##
##  400007: lea rsi, qword ptr [rbx + rax*8]
##          ref!2 = (bvadd (_ bv0 64) (bvadd (_ bv67890 64) (bvmul ((_ extract 63 0) ref!0) (_ bv8 64)))) ; LEA operation
##          ref!3 = (_ bv4194315 64) ; Program Counter
##
##  40000b: lea rsi, dword ptr [ebx + eax*8 + 0xa]
##          ref!4 = ((_ zero_extend 32) (bvadd (_ bv10 32) (bvadd (_ bv67890 32) (bvmul ((_ extract 31 0) ref!0) (_ bv8 32))))) ; LEA operation
##          ref!5 = (_ bv4194321 64) ; Program Counter
##
##  400011: pmovmskb edx, xmm1
##          ref!6 = ((_ zero_extend 32) ((_ zero_extend 16) (concat ((_ extract 127 127) (_ bv0 128)) ((_ extract 119 119) (_ bv0 128)) ((_ extract 111 111) (_ bv0 128)) ((_ extract 103 103) (_ bv0 128)) ((_ extract 95 95) (_ bv0 128)) ((_ extract 87 87) (_ bv0 128)) ((_ extract 79 79) (_ bv0 128)) ((_ extract 71 71) (_ bv0 128)) ((_ extract 63 63) (_ bv0 128)) ((_ extract 55 55) (_ bv0 128)) ((_ extract 47 47) (_ bv0 128)) ((_ extract 39 39) (_ bv0 128)) ((_ extract 31 31) (_ bv0 128)) ((_ extract 23 23) (_ bv0 128)) ((_ extract 15 15) (_ bv0 128)) ((_ extract 7 7) (_ bv0 128))))) ; PMOVMSKB operation
##          ref!7 = (_ bv4194325 64) ; Program Counter
##
##  400015: mov eax, edx
##          ref!8 = ((_ zero_extend 32) ((_ extract 31 0) ref!6)) ; MOV operation
##          ref!9 = (_ bv4194327 64) ; Program Counter
##
##  400017: xor ah, 0x99
##          ref!10 = (concat ((_ extract 63 16) ((_ extract 63 0) ref!8)) (concat (bvxor ((_ extract 15 8) ref!8) (_ bv153 8)) ((_ extract 7 0) ((_ extract 63 0) ref!8)))) ; XOR operation
##          ref!11 = (_ bv0 1) ; Clears carry flag
##          ref!12 = (_ bv0 1) ; Clears overflow flag
##          ref!13 = (bvxor (bvxor (bvxor (bvxor (bvxor (bvxor (bvxor (bvxor (_ bv1 1) ((_ extract 0 0) (bvlshr ((_ extract 15 8) ref!10) (_ bv0 8)))) ((_ extract 0 0) (bvlshr ((_ extract 15 8) ref!10) (_ bv1 8)))) ((_ extract 0 0) (bvlshr ((_ extract 15 8) ref!10) (_ bv2 8)))) ((_ extract 0 0) (bvlshr ((_ extract 15 8) ref!10) (_ bv3 8)))) ((_ extract 0 0) (bvlshr ((_ extract 15 8) ref!10) (_ bv4 8)))) ((_ extract 0 0) (bvlshr ((_ extract 15 8) ref!10) (_ bv5 8)))) ((_ extract 0 0) (bvlshr ((_ extract 15 8) ref!10) (_ bv6 8)))) ((_ extract 0 0) (bvlshr ((_ extract 15 8) ref!10) (_ bv7 8)))) ; Parity flag
##          ref!14 = ((_ extract 15 15) ref!10) ; Sign flag
##          ref!15 = (ite (= ((_ extract 15 8) ref!10) (_ bv0 8)) (_ bv1 1) (_ bv0 1)) ; Zero flag
##          ref!16 = (_ bv4194330 64) ; Program Counter
##

import  sys
from    triton import *


code = [
    (0x400000, "\x48\x8b\x05\xb8\x13\x00\x00"), # mov        rax, QWORD PTR [rip+0x13b8]
    (0x400007, "\x48\x8d\x34\xc3"),             # lea        rsi, [rbx+rax*8]
    (0x40000b, "\x67\x48\x8D\x74\xC3\x0A"),     # lea        rsi, [ebx+eax*8+0xa]
    (0x400011, "\x66\x0F\xD7\xD1"),             # pmovmskb   edx, xmm1
    (0x400015, "\x89\xd0"),                     # mov        eax, edx
    (0x400017, "\x80\xf4\x99"),                 # xor        ah, 0x99
    (0x40001a, "\xC5\xFD\x6F\xCA"),             # vmovdqa    ymm1, ymm2
]



if __name__ == '__main__':

    #Set the arch
    setArchitecture(ARCH.X86_64)

    for (addr, opcodes) in code:
        # Build an instruction
        inst = Instruction()

        # Setup opcodes
        inst.setOpcodes(opcodes)

        # Setup Address
        inst.setAddress(addr)

        # optional - Update register state
        inst.updateContext(Register(REG.RAX,  12345));
        inst.updateContext(Register(REG.RBX,  67890));

        # optional - Add memory access <addr, size, content>
        inst.updateContext(MemoryAccess(0x66666666, 8, 0x31003200330034));

        # Process everything
        processing(inst)

        # Display instruction
        print inst

        # Display symbolic expressions
        for expr in inst.getSymbolicExpressions():
            print '\t', expr

        print

    sys.exit(0)

File: src/examples/cpp/ir.cpp

#include <iostream>
#include <triton/api.hpp>
#include <triton/x86Specifications.hpp>

using namespace triton;
using namespace triton::arch;


struct op {
  unsigned int    addr;
  unsigned char*  inst;
  unsigned int    size;
};

struct op trace[] = {
  {0x400000, (unsigned char *)"\x48\x8b\x05\xb8\x13\x00\x00", 7}, /* mov        rax, QWORD PTR [rip+0x13b8] */
  {0x400007, (unsigned char *)"\x48\x8d\x34\xc3",             4}, /* lea        rsi, [rbx+rax*8]            */
  {0x40000b, (unsigned char *)"\x67\x48\x8D\x74\xC3\x0A",     6}, /* lea        rsi, [ebx+eax*8+0xa]        */
  {0x40000b, (unsigned char *)"\x67\x8D\x74\xC3\x0A",         5}, /* lea        esi, [ebx+eax*8+0xa]        */
  {0x40000b, (unsigned char *)"\x48\x8D\x74\xDB\x0A",         5}, /* lea        rsi, [rbx+rax*8+0xa]        */
  {0x40000b, (unsigned char *)"\x48\x8D\x74\xC3\x0A",         5}, /* lea        rsi, [rbx+rax*8+0xa]        */
  {0x40000b, (unsigned char *)"\x48\x8D\x73\x0A",             4}, /* lea        rsi, [rbx+0xa]              */
  {0x400011, (unsigned char *)"\x66\x0F\xD7\xD1",             4}, /* pmovmskb   edx, xmm1                   */
  {0x400015, (unsigned char *)"\x89\xd0",                     2}, /* mov        eax, edx                    */
  {0x400017, (unsigned char *)"\x80\xf4\x99",                 3}, /* xor        ah, 0x99                    */
  {0x40001a, (unsigned char *)"\x48\x31\xc0",                 3}, /* xor        rax, rax                    */
  {0x40001d, (unsigned char *)"\x80\x30\x99",                 3}, /* xor        byte ptr [rax], 0x99        */
  {0x400020, (unsigned char *)"\x80\x30\x99",                 3}, /* xor        byte ptr [rax], 0x99        */
  {0x400023, (unsigned char *)"\x0F\x87\x00\x00\x00\x00",     6}, /* ja         11                          */
  {0x0,      nullptr,                                         0}
};


int main(int ac, const char **av) {

  /* Set the arch */
  api.setArchitecture(ARCH_X86_64);

  for (unsigned int i = 0; trace[i].inst; i++) {
    /* Build an instruction */
    Instruction inst;

    /* Setup opcodes */
    inst.setOpcodes(trace[i].inst, trace[i].size);

    /* optional - Setup address */
    inst.setAddress(trace[i].addr);

    /* Process everything */
    api.processing(inst);

    std::cout << inst << std::endl;
    for (unsigned int op_index = 0; op_index != inst.operands.size(); op_index++) {
      std::cout << "\tOperand " << op_index << ": " << inst.operands[op_index] << std::endl;
      if (inst.operands[op_index].getType() == OP_MEM) {
        std::cout << "\t   base  : " << inst.operands[op_index].getMemory().getBaseRegister() << std::endl;
        std::cout << "\t   index : " << inst.operands[op_index].getMemory().getIndexRegister() << std::endl;
        std::cout << "\t   disp  : " << inst.operands[op_index].getMemory().getDisplacement() << std::endl;
        std::cout << "\t   scale : " << inst.operands[op_index].getMemory().getScale() << std::endl;
      }
    }

    std::cout << "\t-------" << std::endl;

    for (unsigned int exp_index = 0; exp_index != inst.symbolicExpressions.size(); exp_index++) {
      auto expr = inst.symbolicExpressions[exp_index];
      std::cout << "\tSymExpr " << exp_index << ": " << expr << std::endl;
    }

    std::cout << std::endl << std::endl;
  }

  return 0;
}

